### ISSUE TITLE
Question: Merge Request ID and checklist

### ISSUE THREAD

**Issue by passeris** [melexis/sphinx-traceability-extension#272]
How does the plug-in know which merge request ID to use when generating the traceability table? 

For example, suppose we have in one MR 1/3 items checked, but in another MR, we have 3/3 items checked.

**Comment by passeris** [comment]
I don't quite understand the concept then. Can you provide an example using the checklist in the github [example documents](https://melexis.github.io/sphinx-traceability-extension/meeting_notes.html)? What parts are variable here? Using the example, it seems one has to:

1. change this line in the conf.py: `'merge_request_id': config(MERGE_REQUEST_ID, '121'),  # comma-separated if more than one`
2. Call the makefile with the `MERGE_REQUEST_ID` and the number.

I assume that `MERGE_REQUEST_ID` is constant and does not represent some variable information.

**Comment by Letme** [comment]
`MERGE_REQUEST_ID` is not constant. It needs to be changed before each release to point to the Merge request where checklist for release is checked.

So for example you are preparing release A. You open a merge request (number 213), where you use checklist template. In that merge request you for example bump version to release A and change the MERGE_REQUEST_ID (makefile variable, or variable in CI) to the 213 (id number of the Merge request). Then when you merge the merge request to the master and tag the release, the documentation will be built with `MERGE_REQUEST_ID=213` and checklist data will be taken from that merge request.

When you prepare release B, you again branch out to prepare a release B with checklist template in merge request (id 513). Then you again modify the value of the environment variable (MERGE_REQUEST_ID), either in conf.py or in the Makefile (or call of makefile) and then when merged, the master and tag will take the checklist from the merge request 513.

`config(MERGE_REQUEST_ID)` is from Python Decouple and it is used to load environment variables (or from `.env` file) to define secrets which you do not want to expose in scripts (or you want to externally modify). So you can also just change the `CHECKLIST_MR_ID` in `.env` file if you use that.

**Comment by passeris** [comment]
I seem to have configured my conf.py properly. However, I get no feedback whether it worked or not. For example, in my access token of my .env file, if I introduce a mistake on purpose, the build still works. I am building locally on my machine to see if it contacts the server and gets the results. I am also trying this in my current MR, but nothing is being filled in the checklist. Is there a way to verify that this conf.py indeed is contacting the server?

**Comment by passeris** [comment]
`traceability_checklist = {

  \# 'private_token': config('PRIVATE_TOKEN', ''),
    'api_host_name': config('API_HOST_NAME', 'https://git.swf.daimler.com/api/v4'),
    'project_id': config('PROJECT_ID', 'apricot/apricot-architecture'),
    'merge_request_id': config('MERGE_REQUEST_ID', '879, 850'),
    'checklist_item_regex': r"[A-Z\d_\-]+",
    'attribute_name': config('ATTRIBUTE_NAME', 'checked'),
    'attribute_to_str': config('ATTRIBUTE_TO_STRING', 'Answer'),
    'attribute_values': config('ATTRIBUTE_VALUES', 'yes,no'),
}
`

**Comment by passeris** [comment]
I see, it seems that my api_host_name does not have gitlab in it, but rather `git`. That might be the issue.

**Comment by JasperCraeghs** [comment]
This is the issue for sure. I didn't anticipate an endpoint that does not contain either `gitlab` or `github`. Which one of the two do you need to access? Other options are not supported (yet).

**Comment by passeris** [comment]
when I change to gitlab, and paste it into a browser, I see that it automatically resolves to git. So it seems to be valid with `gitlab` also. However when I change it to `gitlab` I get that strange `builder-inited` exception.

**Comment by passeris** [comment]
> We are using GitLab. When I change to gitlab, and paste it into a browser, I see that it automatically resolves to git. So it seems to be valid with `gitlab` also. However when I change it to `gitlab` I get that strange `builder-inited` exception.

**CORRECTION**, the address does **not** resolve to our GitLab server. Instead it backs out to a central page where we access other tools.

**Comment by JasperCraeghs** [comment]
Try `apricot%2Fapricot-architecture` instead of `apricot/apricot-architecture` or use the project's iid.

See https://docs.gitlab.com/ee/api/index.html#namespaced-path-encoding

**Comment by passeris** [comment]
That worked perfectly! Here is the output. Now the warnings I can take care of in the documents themselves:

```
pickling environment... done
checking consistency... WARNING: List item 'C' in merge/pull request 850 is not defined as a checklist-item.
WARNING: List item 'M' in merge/pull request 850 is not defined as a checklist-item.
WARNING: List item 'L' in merge/pull request 850 is not defined as a checklist-item.
WARNING: List item 'S' in merge/pull request 850 is not defined as a checklist-item.
WARNING: List item 'D' in merge/pull request 850 is not defined as a checklist-item.
WARNING: List item 'P' in merge/pull request 850 is not defined as a checklist-item.
WARNING: List item 'T' in merge/pull request 850 is not defined as a checklist-item.
WARNING: List item 'F' in merge/pull request 850 is not defined as a checklist-item.
WARNING: List item 'I' in merge/pull request 850 is not defined as a checklist-item.
WARNING: List item 'A' in merge/pull request 850 is not defined as a checklist-item.
```

**Comment by Letme** [comment]
Ok, we will keep this issue open to architect something where you can explicitly determine gitlab or github API usage and override the automatic detection we have been using. Would you care to submit a patch?

**Comment by passeris** [comment]
I went ahead and created a pull request where I added an extra elseif in the function that checks the api. See https://github.com/melexis/sphinx-traceability-extension/pull/274 for more details.

**Comment by Letme** [comment]
The feature is released in [v8.7.0](https://github.com/melexis/sphinx-traceability-extension/releases/tag/v8.7.0)

Hope it works as expected and that documentation is now a bit clearer. :sweat_smile:
